clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -1614,14 +1614,39 @@ Decl *TemplateDeclInstantiator::VisitVarDecl(VarDecl *D,
   // Only add this if we aren't instantiating a variable template.  We'll end up
   // adding the VarTemplateSpecializationDecl later.
   if (!InstantiatingVarTemplate) {
-    SemaRef.addSyclVarDecl(Var);
+    if (SemaRef.getLangOpts().SYCLIsDevice &&
+        SemaRef.isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(
+            Var->getType())) {
+      if (!Var->hasGlobalStorage() || Var->isLocalVarDeclOrParm()) {
```

Comment:
There is a check for `!Var->hasGlobalStorage()` and `Var->isLocalVarDeclOrParm()` at the same time? It seems that  `!Var->hasGlobalStorage()` is enough to catch automatic local variable, and local static variable will be caught later in `if (Var->isStaticLocal()) {`.
Removing of `Var->isLocalVarDeclOrParm()` condition will help to avoid double diagnosing, I think.

Reply:
You're right, and that was indeed my course of action. 
I also changed `Var->getAccess() == AS_private` to `Var->getAccess()!=AS_public` 
Thank you.